cmd/precise-code-intel-bundle-manager/internal/database/database.go

```diff
@@ -0,0 +1,460 @@
+type ErrMalformedBundle struct {
+	Filename string // the filename of the malformed SQLite
+	Name     string // the type of value key should contain
+	Key      string // the missing key
+}
+
+func (e ErrMalformedBundle) Error() string {
+	return fmt.Sprintf("malformed bundle: unknown %s %s", e.Name, e.Key)
+}
+
+// BundleMeta represents the values in the meta row.
+type BundleMeta struct {
+	// LSIFVersion is the version of the original index.
+	LSIFVersion string
+	// SourcegraphVersion is the version of the worker that processed the bundle.
+	SourcegraphVersion string
+	// NumResultChunks is the number of rows in the resultChunks table.
+	NumResultChunks int
+}
+
+// ReadMeta reads the first row from the meta table of the given database.
+func ReadMeta(db *sqlx.DB) (BundleMeta, error) {
+	var rows []struct {
+		ID                 int    `db:"id"`
+		LSIFVersion        string `db:"lsifVersion"`
+		SourcegraphVersion string `db:"sourcegraphVersion"`
+		NumResultChunks    int    `db:"numResultChunks"`
+	}
+
+	if err := db.Select(&rows, "SELECT * FROM meta LIMIT 1"); err != nil {
+		return BundleMeta{}, err
+	}
+
+	if len(rows) == 0 {
+		return BundleMeta{}, fmt.Errorf("no rows in meta table")
+	}
+
+	return BundleMeta{
+		LSIFVersion:        rows[0].LSIFVersion,
+		SourcegraphVersion: rows[0].SourcegraphVersion,
+		NumResultChunks:    rows[0].NumResultChunks,
+	}, nil
+}
+
+// OpenDatabase opens a handle to th SQLIte file at the given path.
+func OpenDatabase(filename string, documentDataCache *DocumentDataCache, resultChunkDataCache *ResultChunkDataCache) (*Database, error) {
+	// TODO - What is the behavior if the db is missing? Should we stat first or clean up after?
+	db, err := sqlx.Open("sqlite3_with_pcre", filename)
+	if err != nil {
+		return nil, err
+	}
+
+	meta, err := ReadMeta(db)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		db:                   db,
+		filename:             filename,
+		documentDataCache:    documentDataCache,
+		resultChunkDataCache: resultChunkDataCache,
+		numResultChunks:      meta.NumResultChunks,
+	}, nil
+}
+
+// Close closes the underlyign SQLite handle.
+func (db *Database) Close() error {
+	return db.db.Close()
+}
+
+// Exists determines if the path exists in the database.
+func (db *Database) Exists(path string) (bool, error) {
+	_, exists, err := db.getDocumentData(path)
+	return exists, err
+}
+
+// Definitions returns the set of locations defining the symbol at the given position.
+func (db *Database) Definitions(path string, line, character int) ([]Location, error) {
+	_, ranges, exists, err := db.getRangeByPosition(path, line, character)
+	if err != nil || !exists {
+		return nil, err
+	}
+
+	for _, r := range ranges {
+		if r.DefinitionResultID == "" {
+			continue
+		}
+
+		definitionResults, err := db.getResultByID(r.DefinitionResultID)
+		if err != nil {
+			return nil, err
+		}
+
+		return db.convertRangesToLocations(definitionResults)
+	}
+
+	return nil, nil
```

Comment:
Is the following more semantically correct? i.e. there is no error, and just so happens there is also zero result.
```suggestion
	return []Location{}, nil
```

Reply:
Aren't the semantics equivalent? An empty slice and a nil slice should behave the same.

Reply:
Updated either way.